enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,264 @@
+
+### Open Questions [optional]
+
+- Is `cluster-hosted-net-services-operator` an acceptable name? 
+- In which namespace the new operator components should be hosted?
+
+### current implmentation 
+In current implementation, the [networking infra componenets](https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md) run as static pods through the MCO.
+
+### Suggested design
+
+To support early clustering requirements Keepalived will continue running as static pods through MCO, additionally, a new dnsmasq service (also deployed through MCO) will run in each node while HAProxy, CoreDNS-MDNS and MDNS-publisher will run as daemonset pods through the new operator.
+
+The following [diagram](https://docs.google.com/drawings/d/1a8K16vcaBB6SIF0bN8DaTj4ltjBxVsVNjJN5NCd-U0A/edit?usp=sharing) reflects the suggested changes.
+
+#### Self hosted DNS services
+- The self hosted DNS functionality should be composed of the following components in each node:
+    - A NetworkManager dispatcher script used to configure resolv.conf to point at the node's public IP address, see [self hosted DNS design](https://github.com/openshift/installer/blob/f6bb83784a835e1f4955ef41916793c0948fd4a1/docs/design/baremetal/networking-infrastructure.md#internal-dns) for more details.
+    - Systemd service running dnsmasq
+    - CoreDNS-mDNS pod
+- The CHNSO will be responsible only for the deployment of the CoreDNS pods while the NetworkManager dispatcher script and dnsmasq Systemd service will run through MCO.
+- To avoid port conflict, Dnsmasq service should listen on DNS well-known port (53) and CoreDNS-MDNS should listen on other port (e.g: 5353). 
+- The new Dnsmasq service should be responsible to resolve 'api-int' (to API-VIP value provided in install-config) hostname when CoreDNS-MDNS pod isn't running on the node.
+- CoreDNS-MDNS should register himself as an upstream server in dnsmasq [via dbus interface](https://github.com/imp/dnsmasq/blob/master/dbus/DBus-interface#L51) for ClusterName.DomainName
+ 
+#### Self hosted API Loadbalancer
+- The self hosted API loadbalancer will run similarly to the current mode with one difference: HAProxy pods will run by CHSNO as DaemonSet pods instead of the static pods.
+- API should be available even when HAProxy pods don't run.
+- Additional functionality aspects of the self-hosted Loadbalancer maybe further analyzed upon need in a separate enhancement. 
+
+
+#### Master node shutdown
+In order not to break the [graceful termination process of kube-api](https://github.com/openshift/installer/blob/master/docs/dev/kube-apiserver-health-check.md) the HAProxy pod should be stopped at the same time or after the kube-api static pod.
+
+Using [K8S graceful node shutdown in kubelet](https://github.com/kubernetes/kubernetes/pull/96129) it should be possible to define that HAProxy pod should be terminated after kube-api, by setting HAProxy pod a higher priority than the kube-api static pod.
```

Comment:
@smarterclayton even with this in place, the move of the keepalive VIP to a new node during reboot will disrupt all running requests. This conflicts with our goal of no distruption during rolling deployments. What should we do about this?

cc @squeed

Reply:
If I understand this right though, this issue is pre-existing on bare metal as we had keepalived before already. So it's not a blocker for this enhancement IMO. But this design must set our expectations regarding API outage in CI with bare metal / UPI.

Reply:
That's correct. this enhancement doesn't really make anything worse comparing to the existing implementation. On the contrary, it provides a way to mitigate some of the existing problems by providing an alternative (e.g: ability to use an external LB instead of the self-hosted)